rfqservice/rfq.go

```diff
@@ -0,0 +1,234 @@
+
+	"github.com/lightninglabs/taproot-assets/fn"
+	msg "github.com/lightninglabs/taproot-assets/rfqmessages"
+)
+
+const (
+	// DefaultTimeout is the default timeout used for context operations.
+	DefaultTimeout = 30 * time.Second
+)
+
+// ManagerCfg is a struct that holds the configuration parameters for the RFQ
+// manager.
+type ManagerCfg struct {
+	// PeerMessagePorter is the peer message porter. This component
+	// provides the RFQ manager with the ability to send and receive raw
+	// peer messages.
+	PeerMessagePorter PeerMessagePorter
+}
+
+// Manager is a struct that handles the request for quote (RFQ) system.
+type Manager struct {
+	startOnce sync.Once
+	stopOnce  sync.Once
+
+	// cfg holds the configuration parameters for the RFQ manager.
+	cfg ManagerCfg
+
+	// rfqStreamHandle is the RFQ stream handler. This subsystem handles
+	// incoming and outgoing peer RFQ stream messages.
+	rfqStreamHandle *StreamHandler
+
+	// quoteNegotiator is the RFQ quote negotiator. This subsystem
+	// determines whether a quote is accepted or rejected.
+	quoteNegotiator *QuoteNegotiator
+
+	// ContextGuard provides a wait group and main quit channel that can be
+	// used to create guarded contexts.
+	*fn.ContextGuard
+}
+
+// NewManager creates a new RFQ manager.
+func NewManager(cfg ManagerCfg) (Manager, error) {
+	return Manager{
+		cfg: cfg,
+		ContextGuard: &fn.ContextGuard{
+			DefaultTimeout: DefaultTimeout,
+			Quit:           make(chan struct{}),
+		},
+	}, nil
+}
+
+// Start attempts to start a new RFQ manager.
+func (m *Manager) Start() error {
+	var startErr error
+	m.startOnce.Do(func() {
+		ctx, cancel := m.WithCtxQuitNoTimeout()
+
+		log.Info("Initializing RFQ subsystems")
+		err := m.startSubsystems(ctx)
+		if err != nil {
+			startErr = err
+			return
+		}
+
+		// Start the manager's main event loop in a separate goroutine.
+		m.Wg.Add(1)
+		go func() {
+			defer func() {
+				m.Wg.Done()
+
+				// Attempt to stop all subsystems if the main
+				// event loop exits.
+				err = m.stopSubsystems()
+				if err != nil {
+					log.Errorf("Error stopping RFQ "+
+						"subsystems: %v", err)
+				}
+
+				// The context can now be cancelled as all
+				// dependant components have been stopped.
+				defer cancel()
+			}()
+
+			log.Info("Starting RFQ manager main event loop")
+			err = m.mainEventLoop()
+			if err != nil {
+				startErr = err
+				return
+			}
+		}()
+	})
+	return startErr
+}
+
+// Stop attempts to stop the RFQ manager.
+func (m *Manager) Stop() error {
+	var stopErr error
+
+	m.stopOnce.Do(func() {
+		log.Info("Stopping RFQ system")
+		stopErr = m.stopSubsystems()
+	})
+
+	return stopErr
+}
+
+// startSubsystems starts the RFQ subsystems.
+func (m *Manager) startSubsystems(ctx context.Context) error {
+	var err error
+
+	// Initialise the RFQ raw message stream handler.
+	m.Wg.Add(1)
+	go func() {
+		defer m.Wg.Done()
+		m.rfqStreamHandle, err = NewStreamHandler(
+			ctx, m.cfg.PeerMessagePorter,
+		)
+	}()
+	if err != nil {
```

Comment:
This `err` will always be `nil` here, since the goroutine won't be started immediately.
I guess my question would be: Can these two goroutines here (and the one for `mainEventLoop`) not all be a single event loop?

If not, I would suggest to turn the stream handler and quote negotiator into a fully startable/stoppable independent system that does take care of its own goroutine handling. Otherwise things can get very hard to understand very quickly.

Reply:
> turn the stream handler and quote negotiator into a fully startable/stoppable independent system

I'm going for this. Thanks for taking a look!